### DISCUSSION TITLE
Upgrade Issues: The CSIDriver "csi.hetzner.cloud" is invalid

### DISCUSSION THREAD

**Discussion by snarkipus** [mysticaltech/terraform-hcloud-kube-hetzner#758] [Q&A] [answered]
Working through an upgrade since it's been a while ... successfully got the snapshots built, and everything validates.

`terraform apply` fails at:
```terminal
Terraform will perform the following actions:

  # module.kube-hetzner.data.remote_file.kustomization_backup will be read during apply
...
Plan: 2 to add, 0 to change, 1 to destroy.
module.kube-hetzner.null_resource.kustomization: Destroying... [id=437545420]
module.kube-hetzner.null_resource.kustomization: Destruction complete after 0s
module.kube-hetzner.null_resource.kustomization: Creating...
module.kube-hetzner.null_resource.kustomization: Provisioning with 'file'...
...
module.kube-hetzner.null_resource.kustomization (remote-exec): helmchart.helm.cattle.io/longhorn unchanged
module.kube-hetzner.null_resource.kustomization (remote-exec): The CSIDriver "csi.hetzner.cloud" is invalid: spec.fsGroupPolicy: Invalid value: "File": field is immutable
╷
│ Error: remote-exec provisioner error
│ 
│   with module.kube-hetzner.null_resource.kustomization,
│   on .terraform/modules/kube-hetzner/init.tf line 285, in resource "null_resource" "kustomization":
│  285:   provisioner "remote-exec" {
│ 
│ error executing "/tmp/terraform_1198597371.sh": Process exited with status 1
╵
```

**Comment by mysticaltech** [comment, marked as answer]
@snarkipus FYI https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/742#issuecomment-1518694952

**Reply by snarkipus** [reply]
Thanks @mysticaltech ... that resolved it. Now I'm just wrestling with a stuck helm upgrade for longhorn.

```terminal
Upgrading longhorn
+ helm_v3 upgrade --namespace longhorn-system --repo https://charts.longhorn.io longhorn longhorn --values /config/values-01_HelmChart.yaml
Error: UPGRADE FAILED: rendered manifests contain a resource that already exists. Unable to continue with update: CustomResourceDefinition "backingimagedatasources.longhorn.io" in namespace "" exists and cannot be imported into the current release: invalid ownership  metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "longhorn": current value is "longhorn-crd"
```

**Reply by mysticaltech** [reply]
@snarkipus Get to know helmchartconfig definitions! https://docs.k3s.io/helm

Also, probably interesting to have a look at https://github.com/k3s-io/helm-controller/issues?q=upgrade

Basically to upgrade, you should use a helmchartconfig pointing to your new desired settings, or just edit the helmchart directly. FYI those are CRDs, so you can run `kubectl get helmchart -A`, and same for helmchartconfig, this might helm clarify the matter. Also you can grep the output of kubectl get crds for helm.

**Reply by snarkipus** [reply]
@mysticaltech Thanks - I'll take a look. I just thought you might want to be aware since the helm upgrade was kicked off by upgrading the base install (i.e. I didn't attempt it directly, it just happened as a side effect).

I'm assuming this could happen to anybody that enabled longhorn via `enable_longhorn=true` in their config.

**Reply by mysticaltech** [reply]
@snarkipus Interesting! So many moving variables, but we're getting there, thanks for the info. And PR to the docs or code are always welcome. We need all the help we can get! 🙏
